### PR TITLE
Adding tests CI 

### PR DESCRIPTION
**Pull request by NadavClassiq**
### Some notes

- [ ] Please make sure that you placed the files in an appropriate folder
- [ ] And that the files have indicative names.

- [ ] Please note that Classiq runs automatic code linting, which may minorly alter some files.
  - [ ] If you're familiar with `pre-commit`, you may run `pre-commit install`, and then at each commit, your files will be altered in a similar way

### CODE REVIEW

File: .github/workflows/test-CI.yml [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+          git status
+
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
+
+      # A bunch of if-else. Might move to an action
+      # Decide environment based on the target branch (for both push and PR events)
+      - name: Set environment based on target branch
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:73]: Unite with  "Checkout the pull request's branch" step. Its same logic

Comment (GilClassiq) [reply .github/workflows/test-CI.yml:73]: Assure that the workflow still run from target branch!!!!!

Comment (GilClassiq) [reply .github/workflows/test-CI.yml:73]: Add documentation to this step, make sure all your assumptions are documented:
1. what should run on target branch
2. where the tested code shouold come from
3. what is the different cases - give them meaning

File: .github/workflows/test-CI.yml [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+          if [ "${{ github.event_name }}" == 'pull_request_target' ]; then
+            echo "SHOULD_TEST_ALL_FILES=false" >> $GITHUB_ENV
+          elif [[ "${{ github.event_name }}" == 'workflow_dispatch' || "${{ github.head_ref || github.ref_name }}" == "dev" ]]; then
+            echo "SHOULD_TEST_ALL_FILES=true" >> $GITHUB_ENV
+          fi
+
+      - name: Install dependencies
+        run: |
+          set -e
+          # Pre is needed for Dev pre releases
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:141]: document

File: .github/workflows/test-CI.yml [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          python -m pip install --extra-index-url https://pypi.org/simple --pre -U -r requirements.txt
+          python -m pip install --extra-index-url https://pypi.org/simple -U -r requirements_tests.txt
+
+      - name: Correct directory after manual checkout # Debug step to confirm path
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:145]: why?

Comment (GilClassiq) [reply .github/workflows/test-CI.yml:145]: remove this step

File: tests/utils_for_tests.py [hunk 4]
```diff
@@ -9,10 +9,10 @@ def iterate_notebooks() -> Iterable[str]:
     if os.environ.get("SHOULD_TEST_ALL_FILES", "") == "true":
         notebooks_to_test = _get_all_notebooks()
     else:
-        if os.environ.get("HAS_ANY_IPYNB_CHANGED", "") == "true":
-            notebooks_to_test = os.environ.get("LIST_OF_IPYNB_CHANGED", "").split()
-        else:
-            notebooks_to_test = []
+        # if os.environ.get("HAS_ANY_IPYNB_CHANGED", "") == "true":
```
Comment (GilClassiq) [inline tests/utils_for_tests.py:12]: remove the commented lines

File: .github/workflows/test-CI.yml [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            target_branch="${{ github.event.pull_request.base.ref }}" # Probably should be updated to produce the target branch and not the branch
+          elif [[ "${{ github.event_name }}" == "push" ]]; then
+            echo "were in push"
+            target_branch="$(gh pr view --json baseRefName,headRefName --jq .baseRefName)"
+          else
+            # Use the branch of the push event
+            # todo: verify that dispatch works
+            echo "On dsipatch - WIP"
+            target_branch="${{ github.ref_name }}"
+            echo $target_branch
```
[the remark this answers is not in the document]
Comment (NadavClassiq) [reply .github/workflows/test-CI.yml:84]: Removed + Dispatch trigger was checked and fixed to run all notebooks when triggerred by dispatch

File: .github/workflows/test-CI.yml [hunk 6]
```diff
@@ -1,21 +1,24 @@
 name: Test Library CI
 
+# This workflow always run in the context of Classiq/classiq-library (base repo) but might use a branch from a fork to test its contribution
+# PRs from users (should be from a fork) need to target `main` branch which runs against `prod` environment
+# PRs from classiq team (might be from the base repo) need to target `dev` branch and runs against `dev` environment (nightly)
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:5]: Maintainers team instead of Classiq team

File: .github/workflows/test-CI.yml [hunk 7]
```diff
@@ -30,16 +33,23 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v4
 
+      - name: Print repository name
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:36]: remove

File: .github/workflows/test-CI.yml [hunk 8]
```diff
@@ -49,19 +59,22 @@ jobs:
...
             echo "we're in the right branch. no need to checkout (push)"
+            target_branch="$(gh pr view --json baseRefName,headRefName --jq .baseRefName)" # from old config
           else
             echo "we're not in push/pull/dispatch. error."
             exit 1
           fi
           # debug
           echo "==== after"
           git status
+          echo "Repository Name: ${{ github.repository }}" # Debug - intresting when PR is from a fork
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:77]: Remove the last echo

File: .github/workflows/test-CI.yml [hunk 9]
```diff
@@ -126,13 +120,15 @@ jobs:
         run: |
           echo "Changed notebook files: ${{ steps.changed-files-ipynb.outputs.all_changed_files }}"
 
-      - name: Set changed notebook into environment variables
+      - name: Set changed notebook into environment variables # Could be integrated in logic step
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:123]: move this to the other ugly step

File: .github/workflows/test_scheduled.yml [hunk 10]
```diff
@@ -25,6 +25,9 @@ jobs:
         with:
           python-version: "3.11"
 
+      - name: Debug
```
Comment (GilClassiq) [inline .github/workflows/test_scheduled.yml:28]: remove

File: tests/utils_for_tests.py (hunk 4 above)
Comment (GilClassiq) [inline tests/utils_for_tests.py:12]: REMOVE THE COMMENTSSSSSS

File: .github/workflows/test-CI.yml (hunk 1 above)
Comment (NadavClassiq) [reply .github/workflows/test-CI.yml:87]: Steps were united. 
It's still running in the context of the base repo (not target branch), I tried to 

I tried to explain point #3 in the comment at the beginning of the workflow. If it's not clear enough, I would like you to show me what you mean exactly.

File: .github/workflows/test-CI.yml (hunk 2 above)
Comment (NadavClassiq) [reply .github/workflows/test-CI.yml:127]: I am unsure myself,I probably took it as it is from a script by Zach/ Dor.  As far I understand, it's needed for using pre-release versions. I have clarified it.

File: .github/workflows/test-CI.yml [hunk 11]
```diff
@@ -1,21 +1,24 @@
...
+# PRs from maintainers team (might be from the base repo) need to target `dev` branch and runs against `dev` environment (nightly)
+
 on:
   # Trigger the workflow on push to the specific branch
   push:
     branches:
       - dev
       - main
-      - CAD-22795-restore-changes # Temp
+      - CAD-22795-restore-changes # Remove when merging to main
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:13]: REMOVEEEEEEEEE

File: .github/workflows/test-CI.yml [hunk 12]
```diff
@@ -1,21 +1,24 @@
...
 
   # Trigger the workflow on pull requests targeting the specific branch
   pull_request_target: # Note: `pull_request_target` ensures that the tests run in the context of the `main` branch, not in the user's fork.
-    branches:
+    branches: # It gives access to the secrets as it run in the base repo Classiq/classiq-library so PRs from forks could be tested without exposing it to the user
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:17]: remove this comment

File: .github/workflows/test-CI.yml [hunk 13]
```diff
@@ -1,21 +1,24 @@
...
       - dev
       - main
-      - CAD-22795-restore-changes # Temp
+      - CAD-22795-restore-changes # Remove when merging to main
```
Comment (GilClassiq) [inline .github/workflows/test-CI.yml:20]: Remove

**Review by GilClassiq — APPROVED** [review]

**Pull request merged**